hail/src/main/scala/is/hail/expr/ir/ForwardLets.scala

```diff
@@ -0,0 +1,45 @@
+package is.hail.expr.ir
+
+object ForwardLets {
+  def extract(x: IR): (String, IR, IR) = (x: @unchecked) match {
+    case Let(name, value, body) => (name, value, body)
+    case AggLet(name, value, body) => (name, value, body)
+  }
+
+  private def rewriteIR(ir0: IR): IR = {
+    val UsesAndDefs(uses, _) = ComputeUsesAndDefs(ir0)
+    val nestingDepth = NestingDepth(ir0)
+
+    def rewrite(ir: IR, env: Env[IR]): IR = {
+      ir match {
+        case _: Let | _: AggLet =>
+          val (name, value, body) = extract(ir)
+          val refs = uses.lookup(ir)
+
+          if (value.isInstanceOf[Ref] ||
+            refs.isEmpty ||
+            (refs.size == 1 && nestingDepth.lookup(refs.head) == nestingDepth.lookup(ir)))
+            rewrite(body, env.bind(name -> value))
```

Comment:
I really think we should have two environments here.

Reply:
ugh, yeah. My previous strategy didn't require treating aggenv and env differently, but this definitely does.